### PR TITLE
Generate docker images for jdk17/21 [HZ-2621]

### PR DESCRIPTION
**Pull request by gareth-johnston**

### CODE REVIEW

File: .github/workflows/ee-nlc-snapshot-push.yml [hunk 1]
```diff
@@ -15,6 +15,10 @@ env:
 jobs:
   push:
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        jdk: [ '11', '17', '21' ]
```
Comment (ldziedziul) [inline .github/workflows/ee-nlc-snapshot-push.yml:21]: Do we have approve from PM (or similar person) to add 3 different docker images?

File: .github/workflows/ee-nlc-snapshot-push.yml [hunk 2]
```diff
@@ -85,6 +90,7 @@ jobs:
         run: |
           docker buildx build --push \
             --build-arg HZ_VERSION=${HZ_VERSION} \
+            --build-arg JDK_VERSION=${{ matrix.jdk }} \
             --build-arg HAZELCAST_ZIP_URL=${HAZELCAST_ZIP_URL} \
             --label hazelcast.ee.revision=${{ github.event.inputs.HZ_EE_REVISION }} \
             --tag ${NLC_IMAGE_NAME}:${HZ_VERSION} \
```
Comment (ldziedziul) [inline .github/workflows/ee-nlc-snapshot-push.yml:96]: You need modify tags to have `${NLC_IMAGE_NAME}-jdk${{ matrix.jdk }}:${HZ_VERSION}` for all JDK versions and additionally `${NLC_IMAGE_NAME}-:${HZ_VERSION}` for the default JDK

File: .github/workflows/ee-nlc-tag-push.yml [hunk 3]
```diff
@@ -99,6 +104,7 @@ jobs:
         run: |
           docker buildx build --push \
             --build-arg HZ_VERSION=${HZ_VERSION} \
+            --build-arg JDK_VERSION=${{ matrix.jdk }} \
             --build-arg HAZELCAST_ZIP_URL=${HAZELCAST_ZIP_URL} \
             --tag ${NLC_IMAGE_NAME}:${HZ_VERSION}  \
```
Comment (ldziedziul) [inline .github/workflows/ee-nlc-tag-push.yml:109]: You need modify tags to have `${NLC_IMAGE_NAME}-jdk${{ matrix.jdk }}:${HZ_VERSION}` for all JDK versions and additionally `${NLC_IMAGE_NAME}-:${HZ_VERSION}` for the default JDK

File: .github/workflows/ee_latest_snapshot_push.yml [hunk 4]
```diff
@@ -70,13 +72,26 @@ jobs:
...
           if [[ "${{ github.ref }}" == "refs/heads/master" ]]; then
-            TAGS="${TAGS} --tag ${{ env.DOCKER_ORG }}/hazelcast-enterprise:latest-snapshot${{ matrix.suffix }}"
+            if [[ "${{ matrix.jdk }}" == "$DEFAULT_JDK" ]]; then
+              TAGS="${TAGS} --tag ${{ env.DOCKER_ORG }}/hazelcast-enterprise:latest-snapshot${{ matrix.suffix }}"
+            fi
+            TAGS="${TAGS} --tag ${{ env.DOCKER_ORG }}/hazelcast-enterprise:latest-snapshot${{ matrix.suffix }}-jdk${{ matrix.jdk }}"
+          fi
+          PLATFORMS="linux/arm64,linux/amd64,linux/s390x"
+          # no JDK17 for alpine ppc64le: https://github.com/alpinelinux/aports/commit/b048defcdb41950e18262d382fab8822dc3bd8ee
+          if [[ "${{ matrix.jdk }}" != "17" ]]; then
```
Comment (ldziedziul) [inline .github/workflows/ee_latest_snapshot_push.yml:88]: Have you checked if `ppc64le` in JDK>=17?

File: .github/workflows/oss_latest_snapshot_push.yml [hunk 5]
```diff
@@ -69,13 +73,26 @@ jobs:
...
           if [[ "${{ github.ref }}" == "refs/heads/master" ]]; then
-            TAGS="${TAGS} --tag ${{ env.DOCKER_ORG }}/hazelcast:latest-snapshot${{ matrix.suffix }}"
+            if [[ "${{ matrix.jdk }}" == "$DEFAULT_JDK" ]]; then
+              TAGS="${TAGS} --tag ${{ env.DOCKER_ORG }}/hazelcast:latest-snapshot${{ matrix.suffix }}"
+            fi
+            TAGS="${TAGS} --tag ${{ env.DOCKER_ORG }}/hazelcast:latest-snapshot${{ matrix.suffix }}-jdk${{ matrix.jdk }}"
+          fi
+          PLATFORMS="linux/arm64,linux/amd64"
+          # no JDK17 for alpine ppc64le: https://github.com/alpinelinux/aports/commit/b048defcdb41950e18262d382fab8822dc3bd8ee
+          if [[ "${{ matrix.jdk }}" != "17" ]]; then
```
Comment (ldziedziul) [inline .github/workflows/oss_latest_snapshot_push.yml:89]: Same here, have you checked if `ppc64le` in JDK>=17?

File: .github/workflows/tag_image_push.yml [hunk 6]
```diff
@@ -142,15 +145,27 @@ jobs:
...
+            # Add a tag without JDK version for default JDK
+            if [[ "${{ matrix.jdk }}" == "$DEFAULT_JDK" ]]; then
+                TAGS_ARG="${TAGS_ARG} --tag ${{ env.DOCKER_ORG }}/hazelcast:${tag}${{ matrix.suffix }}"
+            fi
+            # Add a tag with JDK version for all JDKs
+            TAGS_ARG="${TAGS_ARG} --tag ${{ env.DOCKER_ORG }}/hazelcast:${tag}${{ matrix.suffix }}-jdk${{ matrix.jdk }}"
           done
+          PLATFORMS="linux/arm64,linux/amd64"
+          # no JDK17 for alpine ppc64le: https://github.com/alpinelinux/aports/commit/b048defcdb41950e18262d382fab8822dc3bd8ee
+          if [[ "${{ matrix.jdk }}" != "17" ]]; then
```
Comment (ldziedziul) [inline .github/workflows/tag_image_push.yml:160]: Same here: Have you checked if `ppc64le` in JDK>=17?

File: .github/workflows/tag_image_push.yml [hunk 7]
```diff
@@ -159,15 +174,28 @@ jobs:
...
+            if [[ "${{ matrix.jdk }}" == "$DEFAULT_JDK" ]]; then
+                TAGS_ARG="${TAGS_ARG} --tag ${{ env.DOCKER_ORG }}/hazelcast-enterprise:${tag}${{ matrix.suffix }}" 
+            fi
+            # Add a tag with JDK version for all JDKs 
+            TAGS_ARG="${TAGS_ARG} --tag ${{ env.DOCKER_ORG }}/hazelcast-enterprise:${tag}${{ matrix.suffix }}-jdk${{ matrix.jdk }}"
           done
+          
+          PLATFORMS="linux/arm64,linux/amd64,linux/s390x"
+          # no JDK17 for alpine ppc64le: https://github.com/alpinelinux/aports/commit/b048defcdb41950e18262d382fab8822dc3bd8ee
+          if [[ "${{ matrix.jdk }}" != "17" ]]; then
```
Comment (ldziedziul) [inline .github/workflows/tag_image_push.yml:190]: Same here: Have you checked if `ppc64le` in JDK>=17?

**Review by ldziedziul — CHANGES_REQUESTED** [review]

File: .github/workflows/ee-nlc-snapshot-push.yml (hunk 1 above)
Comment (gareth-johnston) [reply .github/workflows/ee-nlc-snapshot-push.yml:21]: From @avtarraikmo - 

For now - lets support 11, 17 and 21 images.

File: .github/workflows/ee-nlc-tag-push.yml [hunk 8]
```diff
@@ -69,10 +73,19 @@ jobs:
...
+          TAGS="--tag ${{ env.NLC_IMAGE_NAME }}-:${{ github.event.inputs.HZ_VERSION }}"
+          fi
+          TAGS="${TAGS} --tag ${{ env.NLC_IMAGE_NAME }}-jdk${{ matrix.jdk }}:${{ github.event.inputs.HZ_VERSION }}"
+          
           docker buildx build --load \
             --build-arg HZ_VERSION=${HZ_VERSION} \
+            --build-arg JDK_VERSION=${{ matrix.jdk }} \
             --build-arg HAZELCAST_ZIP_URL=${HAZELCAST_ZIP_URL} \
-            --tag hazelcast-nlc:test hazelcast-enterprise
+            ${TAGS} \
```
Comment (ldziedziul) [inline .github/workflows/ee-nlc-tag-push.yml:87]: Here we build a test image only, so `hazelcast-nlc:test` is fine

File: .github/workflows/ee-nlc-tag-push.yml [hunk 9]
```diff
@@ -69,10 +73,19 @@ jobs:
...
           docker buildx build --load \
             --build-arg HZ_VERSION=${HZ_VERSION} \
+            --build-arg JDK_VERSION=${{ matrix.jdk }} \
             --build-arg HAZELCAST_ZIP_URL=${HAZELCAST_ZIP_URL} \
-            --tag hazelcast-nlc:test hazelcast-enterprise
+            ${TAGS} \ 
+            hazelcast-enterprise
 
       - name: Run smoke test against EE image
         timeout-minutes: 2
```
Comment (ldziedziul) [inline .github/workflows/ee-nlc-tag-push.yml:91]: Use jdk variable in `Store docker logs as artifact` and `Get docker logs`

File: .github/workflows/ee_latest_snapshot_push.yml [hunk 10]
```diff
@@ -39,6 +40,7 @@ jobs:
       - name: Build Test EE image
         run: |
           docker buildx build --load \
+            --build-arg JDK_VERSION=${{ matrix.jdk }} \
             --build-arg HZ_VERSION=${{ github.event.inputs.HZ_VERSION }} \
             --build-arg HZ_VARIANT=${{ matrix.variant }} \
             --tag hazelcast-ee:test \
```
Comment (ldziedziul) [inline .github/workflows/ee_latest_snapshot_push.yml:46]: Use jdk variable in `Store docker logs as artifact` and `Get docker logs`

File: .github/workflows/ee_latest_snapshot_push.yml (hunk 4 above)
Comment (ldziedziul) [reply .github/workflows/ee_latest_snapshot_push.yml:88]: it seems it's enabled again: https://github.com/alpinelinux/aports/commit/c3c3cf25a6589dad24ba220c23dba70734edd360

but I wonder should we still support all these architectures?

File: .github/workflows/oss_latest_snapshot_push.yml [hunk 11]
```diff
@@ -41,6 +44,7 @@ jobs:
           docker buildx build --load \
             --build-arg HZ_VERSION=${{ github.event.inputs.HZ_VERSION }} \
             --build-arg HZ_VARIANT=${{ matrix.variant }} \
+            --build-arg JDK_VERSION=${{ matrix.jdk }} \
```
Comment (ldziedziul) [inline .github/workflows/oss_latest_snapshot_push.yml:47]: Use jdk variable in `Store docker logs as artifact` and `Get docker logs`

File: .github/workflows/tag_image_push.yml [hunk 12]
```diff
@@ -32,6 +32,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
+        jdk: [ '11', '17', '21' ]
```
Comment (ldziedziul) [inline .github/workflows/tag_image_push.yml:35]: We should use jdk variable in DOCKER_LOG_FILE_OSS and DOCKER_LOG_FILE_EE

File: .github/workflows/tag_image_push.yml [hunk 13]
```diff
@@ -142,15 +145,27 @@ jobs:
           TAGS_TO_PUSH=$(get_tags_to_push ${{ env.RELEASE_VERSION }})
           echo "TAGS_TO_PUSH=$TAGS_TO_PUSH"
           TAGS_ARG=""
+          DEFAULT_JDK=11  # Set the default JDK version
           for tag in ${TAGS_TO_PUSH[@]}
           do
-             TAGS_ARG="${TAGS_ARG} --tag ${{ env.DOCKER_ORG }}/hazelcast:${tag}${{ matrix.suffix }}"
+            # Add a tag without JDK version for default JDK
```
Comment (ldziedziul) [inline .github/workflows/tag_image_push.yml:151]: Use jdk variable in `Store docker logs as artifact` and `Get docker logs`

File: .github/workflows/tag_image_push_rhel.yml [hunk 14]
```diff
@@ -34,6 +34,9 @@ jobs:
       PROJECT_NAME: hztest-${{ github.run_id }}-${{ github.run_attempt }}
 
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        jdk: [ '11', '17', '21' ]
```
Comment (ldziedziul) [inline .github/workflows/tag_image_push_rhel.yml:39]: use jdk variable in ` Get OpenShift events` and `Store OpenShift events as artifact`

**Review by ldziedziul — CHANGES_REQUESTED** [review]

File: .github/workflows/ee-nlc-snapshot-push.yml (hunk 2 above)
Comment (gareth-johnston) [reply .github/workflows/ee-nlc-snapshot-push.yml:96]: Looks like I added this modification to the test build and not this one, will switch around :)

File: .github/workflows/tag_image_push_rhel.yml (hunk 14 above)
Comment (gareth-johnston) [reply .github/workflows/tag_image_push_rhel.yml:39]: done

File: .github/workflows/tag_image_push.yml (hunk 13 above)
Comment (gareth-johnston) [reply .github/workflows/tag_image_push.yml:151]: done

File: .github/workflows/tag_image_push.yml (hunk 12 above)
Comment (gareth-johnston) [reply .github/workflows/tag_image_push.yml:35]: done

File: .github/workflows/oss_latest_snapshot_push.yml (hunk 11 above)
Comment (gareth-johnston) [reply .github/workflows/oss_latest_snapshot_push.yml:47]: done

File: .github/workflows/ee_latest_snapshot_push.yml (hunk 10 above)
Comment (gareth-johnston) [reply .github/workflows/ee_latest_snapshot_push.yml:46]: done

File: .github/workflows/ee-nlc-tag-push.yml (hunk 9 above)
Comment (gareth-johnston) [reply .github/workflows/ee-nlc-tag-push.yml]: done

File: .github/workflows/ee-nlc-tag-push.yml (hunk 8 above)
Comment (gareth-johnston) [reply .github/workflows/ee-nlc-tag-push.yml]: done

**Review by ldziedziul — APPROVED** [review]

**Comment by ldziedziul** [comment]
I've verified workflows by running them against 5.3.6:

- Build EE NLC snapshot image - https://github.com/hazelcast/hazelcast-docker/actions/runs/7301590528
- Build EE NLC tag image - https://github.com/hazelcast/hazelcast-docker/actions/runs/7321222729
- Build EE RHEL image - https://github.com/hazelcast/hazelcast-docker/actions/runs/7327911635
- Build EE snapshot image - https://github.com/hazelcast/hazelcast-docker/actions/runs/7301467004
- Build OS and EE image -  https://github.com/hazelcast/hazelcast-docker/actions/runs/7301534289
- Build OSS snapshot image - https://github.com/hazelcast/hazelcast-docker/actions/runs/7301145624

**Comment by ldziedziul** [comment]
Fixes https://github.com/hazelcast/hazelcast-docker/issues/459

**Pull request merged**